Add latest and trending builds list to the build list module

canuysal opened the pull request:
### Description of the problem being solved:

- Adds a builds list to the BuildList module that displays latest and trending builds that is fetched from the PoB Archives.
- Adds responsive behaviour to the builds menu.

### Steps taken to verify a working solution:
- Tested fetching errors for connection errors, empty build list, some of the keys are missing and success.
- Display status messages depending on the fetch state.
- Checked cpu usage with the new list.
- Checked for invalid and missing pobb links while importing builds

### Before screenshot:
![Screenshot from 2024-03-17 20-12-36](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/59864253/6ed9402e-5524-4c25-a445-f44ff8e196bf)

### After screenshot:
![Screenshot from 2024-03-20 16-41-21](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/59864253/fa6408e6-f3bc-43d7-9b26-adb4ee36c09c)

Wires77 commented on src/Classes/ArchivesListControl.lua:
```diff
@@ -0,0 +1,502 @@
+			if websiteInfo then
+				buildSites.DownloadBuild(button.build_link, websiteInfo, function(isSuccess, data)
+					if isSuccess then
+						local xmlText = Inflate(common.base64.decode(data:gsub("-", "+"):gsub("_", "/")))
```
Sidenote: this might be nice to extract to an import library of sorts

canuysal replied:
Moved this to a function under Common. Also tried to change the usage in the ImportTab but it's just too custom to use a generic function so I left that class as is.

Wires77 reviewed (changes requested):
Thanks for the submission, this looks pretty interesting!  I took a brief look at the code and the feature and I do like what I've seen so far.  I'll start off by saying I'm usually pretty conservative about website tie-ins like this, for several reasons:

- Styling is easier/more freeform in a browser (e.g. "Open in PoB" button and you can style the list however you want)
- Maintenance in case the API changes
- Potential need to integrate with many build list provider websites and keep those up-to-date
- Security concerns about opening malformed builds or malicious links due to bugs/website neglect

Specific concerns for a build list website include the risk of baiting users into bad builds (especially non-beginner builds, given that lack of context available in a list like this), influencing the meta just based on how builds are sorted and (if multiple sites are implemented) which site is displayed by default, and build rot due to an outdated game version.

If we decide to integrate this, there are some changes that I'd like to see to improve it:

- Make the list website-agnostic.  If we integrate pobarchives support, I can easily see others asking for support for their sites.
  - Dropdown at the top showing the website this list is coming from
  - More generic API
    - Latest vs. Trending might not exist for every site
    - Character name probably isn't needed
    - Mapping file/function needed for each site
- Ascendancy assets can be found in TreeData.  I'd love to be able to use those somehow instead of adding duplicate images.
- There's a scrollbar, but scrolling far enough stops showing builds.
- Styling of Trending vs. Latest is a bit misleading, as they're styled like tabs, but the active-looking tab is actually the inactive one (and is clickable)
- Some display issues when in portrait mode:
![image](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/1209372/40116670-73ea-4bf8-8b35-c602c99780b1)

I haven't taken a fine-toothed comb to the code yet, but what I saw seemed pretty well done, so kudos for that and for a decent style otherwise!

canuysal commented:
@Wires77 Please take another look, requested changes are done, and I also implemented the similar builds list as I promised in the discord.

---
New feature, Similar Builds
---
![similarbuilds](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/59864253/98a86654-4148-41da-9353-c3cc8289655b)

- Similar Builds button is hidden by default unless the build is imported from somewhere.
- In order for this to work, I'm saving the import links into the xml's Import node. We can also use this for features such as "Reset to imported version" button in import tab, or we can convert this into a tree to introduce versioning in PoB. I didn't add those since it's out of scope, but let me know if you are interested and we can discuss it further.
- Once the button is pressed, I make an API call to see if there are similar builds for the current build in the Archives. Other build providers can also implement their own API's and lists for this (See the fixes section, lists are now website agnostic.)
- My implementation can only find similar builds if the current one's root build are already added to the Archives and classified by the algorithm. Since I'm collecting almost every build from the youtube and reddit, I think it should work for many. If the build is not in the Archives, list will just say "No builds found."

---
Fixes
---
- Lists are now website agnostic. Providers can implement their own lists by inheriting ExtBuildListProvider abstract class and adding it the provider list in relevant modules. Please let the PoB Archives stay the default as a courtesy for my efforts.
- API's are now modular. Each provider can list their API's as multiple tabs and implement fetching on their classes.
- Fixed responsiveness issues.
- Fixed the issue where tab buttons not indicating the active one correctly.
- Author names are now hidden, but please uncomment that part in ExtBuildList:Draw function and try it with author names while you are reviewing. I think including author names would be better because people would be more interested if a build is from a reliable youtuber/build maker they know.
- Builds listed in this list, if they are coming from the PoB Archives, will be filtered for edits from now on. So if users edit their builds in any way in the site, it wont be listed to prevent title abuse and injection.
- I didn't remove the new ascendancy portraits, I tried the ones in TreeData but they are made for TreeView, and I think they look bad as 80x80 portraits, also indistinguishable. I can still use the ones in TreeData or move new ones to an appropriate place, please let me know.

![wideangleascendancies1](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/59864253/ad0e8f8a-1a25-408e-8bbf-225d2c01f362)

---
Known issues, need help
---
Please point me in the right direction for these issues, direct fixes by others are also appreciated.
- While in Tree View, similar builds list and the tree share the scrollbar. I'm guessing this can be fixed in PassiveTreeView.
- Scrollbar doesn't snap to the bottom while scrolling and makes multiple round trips up and down.
- In the Builds module, scrollbar doesn't start working until the list is clicked.

![known_issues](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/59864253/1fadcf1e-581e-44fb-9898-9b939bc147bc)


Feel free to add new commits fellow PoB Warriors, I'm happy to share this PR and any help/improvement is appreciated.

Nightblade commented on src/Classes/ExtBuildListProvider.lua:
Please ensure this and all your changed/added code use tab-indents for code-base consistency.

canuysal replied:
Sorry I didn't realize that. It looks like only this file had space indentation somehow.. Fixed.

Wires77 commented on src/Classes/ExtBuildListControl.lua:
```diff
@@ -0,0 +1,551 @@
+
+			end
+
+			-- import button
+			local importButton = {
+				buildLink = build.buildLink,
+				buildName = build.buildName,
+				authorName = build.author,
+				x0 = x,
+				y0 = currentHeight + 6,
+				x1 = x + 47,
+				y1 = currentHeight + 26
+			}
```
Why not just create actual buttons here?  They come with everything you need, like placement, width, height, click handler, etc.  I'm not immediately seeing anything that requires a bespoke solution like this.

canuysal replied:
I thought about implementing a build card and drawing each build as its own control class, that way buttons would be also taken care of with buttoncontrol. But integrating that with scrollbar class would cause issues such as partial drawing so I left as is. Never thought about only converting buttons though, I will try it now.

canuysal replied:
Spent whole day trying to create a better way (card class), then a few hours on implementing just the buttons but its finally fixed! Main reason that made it a hassle was anchoring/preventing overflows while keeping it under controls table.

Wires77 commented on src/Modules/Build.lua:
```diff
@@ -1398,7 +1408,8 @@ function buildMode:OpenSaveAsPopup()
 	controls.label = new("LabelControl", nil, 0, 20, 0, 16, "^7Enter new build name:")
-	controls.edit = new("EditControl", nil, 0, 40, 450, 20, self.dbFileName and self.buildName, nil, "\\/:%*%?\"<>|%c", 100, function(buf)
+	controls.edit = new("EditControl", nil, 0, 40, 450, 20,
+	(self.dbFileName or self.buildName):gsub("[\\/:%*%?\"<>|%c]", "-"), nil, "\\/:%*%?\"<>|%c", 100, function(buf)
```
This causes issues when using "Save As" on a local build:
![image](https://github.com/user-attachments/assets/6471a593-cf1a-4c5d-a434-61f1c2fcd198)

canuysal replied:
Fixed.

Wires77 commented on src/Classes/PopupDialog.lua:
```diff
@@ -55,6 +58,9 @@ function PopupDialogClass:Draw(viewPort)
 	if self.scrollBarFunc then
 		self.scrollBarFunc()
 	end
+	if self.resizeFunc then
+		self.resizeFunc()
+	end
```
It might be more generically useful to check the height and width compared to the viewport/screensize and if it doesn't fit anymore, resize accordingly instead of making each popup implement their own logic for it.

Wires77 commented on src/Modules/Common.lua:
```diff
@@ -921,4 +921,30 @@ function string:matchOrPattern(pattern)
 		end
 	end
 	return false
+end
+
+function ImportBuild(importLink, callback)
+	local urlText = importLink:gsub("^[%s?]+", ""):gsub("[%s?]+$", "") -- Quick Trim
+	if urlText:match("youtube%.com/redirect%?") or urlText:match("google%.com/url%?") then
+		local nested_url = urlText:gsub(".*[?&]q=([^&]+).*", "%1")
+		urlText = UrlDecode(nested_url)
+	end
+	local websiteInfo = nil
+	for j = 1, #buildSites.websiteList do
+		if urlText and urlText:match(buildSites.websiteList[j].matchURL) then
+			websiteInfo = buildSites.websiteList[j]
+		end
+	end
+
+	-- its an import link
+	if websiteInfo then
+		buildSites.DownloadBuild(urlText, websiteInfo, function(isSuccess, data)
+			if isSuccess then
+				callback(Inflate(common.base64.decode(data:gsub("-", "+"):gsub("_", "/"))), urlText)
+			end
+		end)
+	else
+		-- try to decode input buffer
+		callback(Inflate(common.base64.decode(importLink:gsub("-", "+"):gsub("_", "/"))), nil)
+	end
```
If this is being added, we should use it in the ImportTab as well so we don't duplicate code.

Wires77 replied:
Had forgotten we already talked about this here: https://github.com/PathOfBuildingCommunity/PathOfBuilding/pull/7389#discussion_r1612344874

I still think it'd be nice to reuse code, but that can be a separate improvement if we have to.

Wires77 commented on src/Classes/ExtBuildListControl.lua:
```diff
@@ -0,0 +1,551 @@
+	end
+end
+
+function ExtBuildListControlClass:GetAscendancyImageHandle(ascendancy)
+	local image = nil
+	if ascendancy then
+		image = NewImageHandle()
+		image:Load(s_format('Assets/ascendants/%s.jpeg', (ascendancy:gsub("^%l", string.lower))))
+	end
+
+	return image
+end
+
+function ExtBuildListControlClass:GetClassImageHandle(class)
+	local image = nil
+	if class then
+		image = NewImageHandle()
+		image:Load(s_format('Assets/ascendants/%s.jpeg', (class:gsub("^%l", string.lower))))
+	end
+
+	return image
+end
```
This is the same function

canuysal replied:
Yeah in my first class fallback implementation I thought I would add something special for classes but I didn't after all. Fixed into single function now.

Wires77 commented on src/Classes/ExtBuildListControl.lua:
```diff
@@ -0,0 +1,551 @@
+	for _, build in pairs(self.activeListProvider.buildList) do
+		if build.buildName then
+			local image = nil
+			if build.ascendancy then
+				image = self:GetAscendancyImageHandle(build.ascendancy)
+			else if build.class then
+				image = self:GetClassImageHandle(build.class)
+				end
+			end
```
Simplified:

```suggestion
			if build.ascendancy or build.class then
				image = self:GetImageHandle(build.ascendancy or build.class)
			end
```
Also, just for future reference:
```suggestion
                        if build.ascendancy then
				image = self:GetAscendancyImageHandle(build.ascendancy)
			elseif build.class then
				image = self:GetClassImageHandle(build.class)
			end
```

canuysal replied:
Fixed with the previous commit.

Wires77 commented on src/Classes/ExtBuildListControl.lua:
```diff
@@ -0,0 +1,551 @@
+				end
+			end
+
+			if image then
```
Might as well put this in the if block I suggested too

canuysal replied:
Fixed with the previous commits.

Wires77 reviewed (changes requested):
Looks pretty good overall, but I did have some concerns on the code side.  Largest one being around the custom button code, since that seems pretty unmaintainable going forward.

Also keep in mind we don't have a Warden image to use (yet), so this is what that'll look like on league start:
![image](https://github.com/user-attachments/assets/45d36a30-9e94-41e6-ad03-a7cf56f1b2e5)

Nightblade commented on src/Modules/Build.lua:
```diff
@@ -1459,6 +1477,47 @@ function buildMode:OpenSpectreLibrary()
+		}
+	}
+	local width = 600
+    local height = function()
+        return main.screenH * 0.8
+    end
+    local padding = 50
+    controls.similarBuildList = new("ExtBuildListControl", nil, 0, padding, width, height() - 2 * padding, buildProviders)
+    controls.similarBuildList.shown = true
+    controls.similarBuildList.height = function()
+        return height() - 2 * padding
```
Some space indentation here.

canuysal replied:
Fixed.

Wires77 commented on src/Classes/ExtBuildListControl.lua:
```diff
@@ -0,0 +1,445 @@
+				end
+
+				if build.previewLink then
+					local previewButton = new("ButtonControl", nil, 0, 0, 60, 20, "Preview", function()
+							OpenURL(build.previewLink)
+					end)
+					previewButton.x = x + 50
+					previewButton.y = currentHeight - self.controls.scrollBarV.offset
+					t_insert(self.controls, previewButton)
```
```suggestion
					local previewButton = new("ButtonControl", nil, x + 50, currentHeight - self.controls.scrollBarV.offset, 60, 20, "Preview", function()
							OpenURL(build.previewLink)
					end)
					t_insert(self.controls, previewButton)
```

Bellabong commented:
If anything, this should link directly to the build link that Divina scraped from, and not to PoB Archives.

Araxeus commented:
FYI This PR is a hot topic on reddit

https://www.reddit.com/r/pathofexile/comments/1ec10fo/path_of_building_has_been_hijacked_by_a_forprofit/
